internal/generator/ecc/template/sswu-fp.go.tmpl

```diff
@@ -0,0 +1,397 @@
+func sqrtRatio(z *{{.CoordType}}, u *{{.CoordType}}, v *{{.CoordType}}) bool {
+	var tv1 {{.CoordType}}
+	tv1.Square(v)
+	var tv2 {{.CoordType}}
+	tv2.Mul(u, v)
+	tv1.Mul(&tv1, &tv2)
+
+	var y1 {{.CoordType}}
+	{
+	var c1 big.Int
+	c1.SetBytes([]byte { {{ printList (bytes (index .SqrtRatioParams 0) ) }} })
+	y1.Exp(tv1, &c1)
+	}
+
+	y1.Mul(&y1, &tv2)
+
+	var y2 {{.CoordType}}
+	y2.Mul(&y1, &{{.CoordType}}{ {{ words64 (index .SqrtRatioParams 1) }} })
+
+	var tv3 {{.CoordType}}
+	tv3.Square(&y1)
+	tv3.Mul(&tv3, v)
+	isQr := tv3 == *u
+
+	if isQr {
+		*z = y1
+	} else {
+		*z = y2
+	}
+
+	return isQr
+}
+
+{{ end }}
+
+{{ if eq (mod .FieldSizeMod256 8) 5 }}
+
+// sqrtRatio computes the square root of u/v and returns true if u/v was indeed a quadratic residue
+// if not, we get sqrt(Z * u / v). Recall that Z is non-residue
+// Taken from https://datatracker.ietf.org/doc/draft-irtf-cfrg-hash-to-curve/13/ F.2.1.3. q = 5 mod 8
+// The main idea is that since the computation of the square root involves taking large powers of u/v, the inversion of v can be avoided
+func sqrtRatio(z *{{.CoordType}}, u *{{.CoordType}}, v *{{.CoordType}}) bool {
+
+    var tv1, tv2 {{.CoordType}}
+    tv1.Square(v)
+    tv2.Mul(&tv1, v)
+    tv1.Square(&tv1)
+    tv2.Mul(&tv2, u)
+    tv1.Mul(&tv1, &tv2)
+
+    var c1 big.Int
+    c1.SetBytes([]byte { {{ printList (bytes (index .SqrtRatioParams 0) ) }} })
+    var y1 {{.CoordType}}
+    y1.Exp(tv1, &c1)
+    y1.Mul(&y1, &tv2)
+    tv1.Mul(&y1, &{{.CoordType}} { {{words64 (index .SqrtRatioParams 1) }} })
+    tv2.Square(&tv1)
+
+    //Line 10 in std doc
+    tv2.Mul(&tv2, v)
+
+    if tv2 == *u {
+        y1 = tv1
+    }
+
+    tv2.Square(&y1)
+    tv2.Mul(&tv2, v)
+
+    //Line 15
+    isQr := tv2 == *u
+    var y2 {{.CoordType}}
+    y2.Mul(&y1, &{{.CoordType}} { {{words64 (index .SqrtRatioParams 2) }} } )
+    tv1.Mul(&y2, &{{.CoordType}} { {{words64 (index .SqrtRatioParams 1) }} })
+    tv2.Square(&tv1)
+    tv2.Mul(&tv2, v)
+
+    var tv3 {{.CoordType}}
+    //Line 20
+    // Z = {{.Z}}
+    mulByZ(&tv3, u)
+    if tv2 == tv3 {
+        y2 = tv1
+    }
+    if isQr {
+        *z = y1
+    } else {
+        *z = y2
+    }
+    
+    return isQr
+}
+
+{{ end }}
+
+{{ if eq (mod .FieldSizeMod256 8) 1 }}
+// sqrtRatio computes the square root of u/v and returns true if u/v was indeed a quadratic residue
+// if not, we get sqrt(Z * u / v). Recall that Z is non-residue
+// The return value is undefined for u = 0
+// Taken from https://datatracker.ietf.org/doc/draft-irtf-cfrg-hash-to-curve/13/ F.2.1.1. for any field
+// The main idea is that since the computation of the square root involves taking large powers of u/v, the inversion of v can be avoided
+func sqrtRatio(z *{{.CoordType}}, u *{{.CoordType}}, v *{{.CoordType}}) bool {
+   tv1 := {{.CoordType}} { {{words64 (index .SqrtRatioParams 5) }} }
+
+   var exp big.Int
+   exp.SetBytes([]byte { {{ printList (bytes (index .SqrtRatioParams 3) ) }} })
```

Comment:
big.Int SetBytes is not very readable on the generated code, why not set from a hex string value?

Reply:
Reading from hex string is very slow (20% of entire runtime initially). But I can add a comment for readability.